johnnyfive/gmail.py

```diff
@@ -0,0 +1,396 @@
+
+        # Initialize the Gmail connection
+        self.service = setup_gmail(interactive=interactive)
+
+        # Build the container for a multipart MIME message
+        self.message = multipart.MIMEMultipart()
+        self.message['to'] = toaddr if isinstance(toaddr, str) else ','.join(toaddr)
+        self.message['from'] = f"{fromname} <{fromaddr}>" if fromname else fromaddr
+        self.message['subject'] = subject
+
+        # Place the text into the message
+        self.message.attach( text.MIMEText(message_text))
+
+    def add_attachment(self, file):
+        """add_attachment _summary_
+
+        _extended_summary_
+
+        Parameters
+        ----------
+        file : `str`
+            Filename of the attachment
+        """
+        # For the attachment, guess the MIME type for reading it in
+        content_type, encoding = mimetypes.guess_type(file)
+
+        # Set unknown type
+        if content_type is None or encoding is not None:
+            content_type = 'application/octet-stream'
+
+        # Case out the content type
+        main_type, sub_type = content_type.split('/', 1)
+        if main_type == 'text':
+            with open(file, 'rb') as fp:
+                attachment = text.MIMEText(fp.read(), _subtype=sub_type)
+        elif main_type == 'image':
+            with open(file, 'rb') as fp:
+                attachment = image.MIMEImage(fp.read(), _subtype=sub_type)
+        elif main_type == 'audio':
+            with open(file, 'rb') as fp:
+                attachment = audio.MIMEAudio(fp.read(), _subtype=sub_type)
+        else:
+            with open(file, 'rb') as fp:
+                attachment = base.MIMEBase(main_type, sub_type)
+                attachment.set_payload(fp.read())
+
+        # Add the attachment to the email message
+        attachment.add_header('Content-Disposition', 'attachment',
+                           filename=os.path.basename(file))
+        self.message.attach(attachment)
+
+    def send(self):
+        """send Send the GmailMessage
+
+        _extended_summary_
+
+        Returns
+        -------
+        `dict`
+            The sent message object
+        """
+        # Take the message object, and 64-bit encode it properly for sending
+        encoded_message = base64.urlsafe_b64encode(self.message.as_bytes())
+        # The sendable message is a dictionary containing the raw decoded thing
+        sendable_message = {'raw': encoded_message.decode()}
+
+        # Go into a possibly infinite loop!
+        # TODO: Make this NOT an infinite loop
+        while not self.service:
+            self.service = setup_gmail()
+
+        # Try to send the message (API: users.messages.send)
+        try:
+            message = self.service.users().messages().send(userId="me",
+                                            body=sendable_message).execute()
+            #print(f"Message Id: {message['id']}")
+            return message
+        except HttpError as error:
```

Comment:
If self.service is None, there will probably be an attribute error or some other gremlin in here that's worth catching.  You could just catch that error and say "Check gmail credentials" since it implies that ```setup_gmail``` has failed or you could check if self.service is None, but then you're doing the same comparison as in my suggestion and it probably makes sense to refactor a bit to just combine all that jazz